src/devices/badger_water.c

```diff
@@ -0,0 +1,157 @@
+encoding. This code treats 6 bits of the sync sequence as part of a 16 bit preamble.
+
+Once the data has been decoded with the NRZ 6:4 decoding, it has the following format:
+- Device ID: 3 bytes, little-endian. Value as stamped on cover, mod 2^24
+- Device flags: 1 byte. Fields not known
+- Meter reading: 3 bytes, little-endian. Value in gallons for meters with 1-gallon resolution.
+- Status flags: 1 byte. Fields not known
+- CRC: 2 bytes, crc16, polynomial 0x3D65
+
+*/
+
+// Mapping from 6 bits to 4 bits. "3of6" coding used for Mode T
+static uint8_t badger_decode_3of6(uint8_t byte)
+{
+    uint8_t out = 0xFF; // Error
+    //fprintf(stderr,"Decode %0d\n", byte);
+    switch(byte) {
+    case 22:    out = 0x0;  break;  // 0x16
+    case 13:    out = 0x1;  break;  // 0x0D
+    case 14:    out = 0x2;  break;  // 0x0E
+    case 11:    out = 0x3;  break;  // 0x0B
+    case 28:    out = 0x4;  break;  // 0x1C
+    case 25:    out = 0x5;  break;  // 0x19
+    case 26:    out = 0x6;  break;  // 0x1A
+    case 19:    out = 0x7;  break;  // 0x13
+    case 44:    out = 0x8;  break;  // 0x2C
+    case 37:    out = 0x9;  break;  // 0x25
+    case 38:    out = 0xA;  break;  // 0x26
+    case 35:    out = 0xB;  break;  // 0x23
+    case 52:    out = 0xC;  break;  // 0x34
+    case 49:    out = 0xD;  break;  // 0x31
+    case 50:    out = 0xE;  break;  // 0x32
+    case 41:    out = 0xF;  break;  // 0x29
+    default:    break;  // Error
+    }
+    return out;
+}
+
+
+// Decode the DC-free 4:6 encoding
+static int badger_decode_3of6_buffer(uint8_t const *bits, unsigned bit_offset, uint8_t *output)
+{
+    for (unsigned n=0; n<10; ++n) {
+        uint8_t nibble_h = badger_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset) >> 2);
+        uint8_t nibble_l = badger_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset+6) >> 2);
+        if ((nibble_h | nibble_l) > 15) {
+            return 1;
+        }
+        output[n] = (nibble_h << 4) | nibble_l;
+    }
+    return 0;
+}
+
+static int badger_orion_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    static const uint8_t PREAMBLE_ORION[] = {0x54, 0x3D};
+
+    uint8_t data_in[10] = {0}; // Data from Physical layer decoded to bytes
+    data_t *data;
+
+    // Validate package length
+    // The minimum preamble is 16 bits and the payload is 10 4:6 encoded bytes.
+    // There is often a long preamble and a 64 or more bits of tail, so the maximum likely length is longer
+    if (bitbuffer->bits_per_row[0] < (16 + 12 * 10) || bitbuffer->bits_per_row[0] > (128 + 16 + 12 * 10 + 96)) {
+        return DECODE_ABORT_LENGTH;
+    }
+
+    // Find the preamble
+    unsigned bit_offset = bitbuffer_search(bitbuffer, 0, 0, PREAMBLE_ORION, sizeof(PREAMBLE_ORION) * 8);
+    if (bit_offset + 12 * 10 >= bitbuffer->bits_per_row[0]) {
+        return DECODE_ABORT_EARLY;
+    }
+
+    decoder_logf_bitbuffer(decoder, 2, __func__, bitbuffer, "PREAMBLE_ORION: found at: %u", bit_offset);
```

Comment:
`Preamble found at`  should be used, the context for the log is decoder and func automatically.